feat: adds guide for React Native on iOS (#85)
Basis-Theory/developers.basistheory.com: docs/guides/collect/collect-data-with-react-native-on-ios.mdx

import { Alert } from "@site/src/components/shared/Alert";
import { AuthButtons } from "@site/src/components/docs/AuthButtons";

> # Collect Data with React Native on iOS

Reviewer: this is more commentary, although i think we are going to need to invest in using tabs on these pages instead of listing them all in the sidebar soon.

## Getting Started

To get started, you will need a Basis Theory account.

<AuthButtons />

Next you will need a [Public Application](/docs/api/applications#application-types) in order to use [iOS Elements](/docs/sdks/mobile/ios) for your app.

[Click here](https://portal.basistheory.com/applications/create?permissions=token%3Acreate&type=public&name=Collect%20Data%20from%20iOS%20Guide) to create a Public Application or
[login to your Basis Theory account](https://portal.basistheory.com/applications) and create a new application with the following settings:

- Name - Collect Data from iOS Guide
- Application Type - Public
- Permissions - `token:create`

<Alert>
  Save the API Key from the created Public Application as it will be used later
  in this guide to initialize the form.
</Alert>

## Setup the Project

We will create a new React Native application by following the instructions outlined on React Native's environment setup [page](https://reactnative.dev/docs/environment-setup). Ensure you have
the following tabs selected before starting:

1. "React Native CLI Quickstart" near the top of the page
2. "macOS" selected as the Development OS
3. "iOS" selected as the Target OS

> Follow each step on this page down to [Start your application](https://reactnative.dev/docs/environment-setup#step-2-start-your-application) to ensure your environment is set up correctly.

Reviewer 2: This URL does not work as expected. When I clicked on it for the first time it didn't take me to the right section bc the tab `React Native CLI Quickstart` is not selected by default.

Author: yeah.. maybe I should just leave this link off since there doesn't seem to be a way to pre-select the right tabs

## Install the iOS Elements SDK

We will need to install Basis Theory's [iOS Elements SDK](/docs/sdks/mobile/ios), which will render a secure text field for capturing the data.

Open your new React Native project in Xcode by opening `AwesomeProject.xcworkspace` in your project under the `ios` directory.

> With the Xcode window in focus, select File > Add Packages. In the dialog that appears, enter "[https://github.com/Basis-Theory/basistheory-ios](https://github.com/Basis-Theory/basistheory-ios)" in the search field on the top right, then select "AwesomeProject" for Add to Project,

Reviewer: is this like this because it is ReactNative and you can't install it with `pod install`?

Reviewer: is that just true for this ReactNative part or for any iOS app?

Author: it's more complex because it's within React Native. it works fine an a native iOS app

and tokenize its values, respectively.

```objc showLineNumbers title="AwesomeProject/SsnTextElement.m"
#import <Foundation/Foundation.h>
// highlight-start
#import "React/RCTBridgeModule.h"
#import "React/RCTViewManager.h"

@interface RCT_EXTERN_MODULE(SsnTextElement, RCTViewManager)

RCT_EXTERN_METHOD(
                  tokenize: (RCTPromiseResolveBlock) resolve
                  rejecter: (RCTPromiseRejectBlock) reject
                  )
@end
// highlight-end
```

Add a new Swift file through Xcode called `SsnTextElement.swift` under the "AwesomeProject" project in the "AwesomeProject" directory. After going through the prompts for adding your new `SsnTextElement.swift` file, opt into creating a bridging header for your project.
We will need this bridging header in a bit.

Let's import our necessary packages into `SsnTextElement.swift`.

```swift showLineNumbers title="AwesomeProject/SsnTextElement.swift"
import Foundation
// highlight-start
import UIKit
import BasisTheoryElements
// highlight-end
```

> At this point, you may get an error telling you that `BasisTheoryElements` isn't compatible with the minimum iOS deployment set up for the project. In that case, in the Project Navigator tab in Xcode click on the "AwesomeProject" project and under the "AwesomeProject" target,

Reviewer: Do we have a way to make this a "info" or "warning" callout?

select "14.0" for the Minimum Deployment.

> Now we need to make `RCTViewManager` available to be used in Swift. We'll go back to the bridging header created earlier and add an import for the `RCTViewManager` header.

Reviewer: These sentences sound a bit odd? I'm not sure exactly what to change although: 

```suggestion
Back in the bridging header we need to make `RCTViewManager` available in Swfit. To accomplish this, we need to add an import for the `RCTViewManager` header.
```

Author: lol I'm not a fan of this suggestion either though

```objc showLineNumbers title="AwesomeProject/AwesomeProject-Bridging-Header.h"
//
//  Use this file to import your target's public headers that you would like to expose to Swift.
//
// highlight-start
#import "React/RCTViewManager.h"
// highlight-end
```

## Add Your Form Component on iOS

> Next we'll setup our component to our liking. Let's [initiate](/docs/sdks/mobile/ios#initialization) our element, add a [mask](/docs/sdks/mobile/ios/options#mask), and add some [styling](/docs/sdks/mobile/ios#styling-elements) in our `SsnTextElement.swift` file

Reviewer: ```suggestion
Next we'll setup our component to our liking. Let's [initiate](/docs/sdks/mobile/ios#initialization) the TextElementUITextField, add a [mask](/docs/sdks/mobile/ios/options#mask), and add some [styling](/docs/sdks/mobile/ios#styling-elements) in our `SsnTextElement.swift` file
```

with the following code:

```swift showLineNumbers title="AwesomeProject/SsnTextElement.swift"
import Foundation
import UIKit
import BasisTheoryElements

// highlight-start
@objc(SsnElement) class SsnElement: RCTViewManager {
  let ssnTextElement = TextElementUITextField()

  override func view() -> UIView! {
    let regexDigit = try! NSRegularExpression(pattern: "\\d")
    let ssnMask: [Any] = [
      regexDigit,
      regexDigit,
      regexDigit,
      "-",
      regexDigit,
      regexDigit,
      "-",
      regexDigit,
      regexDigit,
      regexDigit,
      regexDigit,
    ]

    try! ssnTextElement.setConfig(options: TextElementOptions(mask: ssnMask))

    ssnTextElement.layer.cornerRadius = 16.0
    ssnTextElement.layer.borderWidth = 1.0
    ssnTextElement.layer.borderColor = UIColor( red: 209/255, green: 209/255, blue: 214/255, alpha: 1.0 ).cgColor
    ssnTextElement.layer.masksToBounds = true
    ssnTextElement.backgroundColor = .white
    ssnTextElement.placeholder = "Enter SSN"
    ssnTextElement.leftView = UIView(frame: CGRect(x: 0, y: 0, width: 20, height: 20))
    ssnTextElement.leftViewMode = .always

    return ssnTextElement
  }
}
// highlight-end
```

> Now let's add our function to tokenize. Notice in this function how we make use of the `resolve` and `rejector` params. These will get used to communicate the results of the tokenize function back to our Javascript.

Reviewer: Should this be a separate step?

```suggestion
Now let's make sure we can tokenize data entered into our `ssnTextElement`. 

In this function, we make use of the `resolve` and `rejector` params. These will are used to communicate the results of the tokenize function back to our Javascript.
```

    ssnTextElement.layer.borderColor = UIColor( red: 209/255, green: 209/255, blue: 214/255, alpha: 1.0 ).cgColor
    ssnTextElement.layer.masksToBounds = true
    ssnTextElement.backgroundColor = .white
    ssnTextElement.placeholder = "Enter SSN"
    ssnTextElement.leftView = UIView(frame: CGRect(x: 0, y: 0, width: 20, height: 20))
    ssnTextElement.leftViewMode = .always

    return ssnTextElement
  }

  // highlight-start
  @objc public func tokenize(
    _ resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: RCTPromiseRejectBlock
  ) -> Void {
    let body = CreateToken(type: "token", data: [
      "ssn": self.ssnTextElement,
    ])

    BasisTheoryElements.createToken(body: body, apiKey: "test_1234567890") {
      data, error in
      guard error == nil else {
        resolve(error!.localizedDescription)
        return
      }

      let encoder = JSONEncoder()
      let json = try! encoder.encode(data)
      resolve(String(data: json, encoding: .utf8)!)
    }
  }
  // highlight-end
}
```

<Alert>
  Be sure to replace <code>test_1234567890</code> with the Public API Key you
  created in the <a href="#getting-started">Getting Started</a> step.
</Alert>

> All that's left is to let React Native know that we need to setup our `SsnTextElement` on the main queue.

Reviewer: ```suggestion
All that's left is setup `SsnTextElement` on the main queue, making ti available to React Native.
```

Reviewer: i'm not sure we do anything, although in the page this sentence kind of blends in.. Maybe there is a way to do sub-steps for this and tokenize?

Author: I feel weird about this suggestion. we need to set up our UITextField on the main queue for iOS processes. I wouldn't feel comfortable saying it's what makes it available in React Native

    ssnTextElement.layer.borderWidth = 1.0
    ssnTextElement.layer.borderColor = UIColor( red: 209/255, green: 209/255, blue: 214/255, alpha: 1.0 ).cgColor
    ssnTextElement.layer.masksToBounds = true
    ssnTextElement.backgroundColor = .white
    ssnTextElement.placeholder = "Enter SSN"
    ssnTextElement.leftView = UIView(frame: CGRect(x: 0, y: 0, width: 20, height: 20))
    ssnTextElement.leftViewMode = .always

    return ssnTextElement
  }

  @objc public func tokenize(
    _ resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: RCTPromiseRejectBlock
  ) -> Void {
    let body = CreateToken(type: "token", data: [
      "ssn": self.ssnTextElement,
    ])

    BasisTheoryElements.createToken(body: body, apiKey: "test_1234567890") {
      data, error in
      guard error == nil else {
        resolve(error!.localizedDescription)
        return
      }

      let encoder = JSONEncoder()
      let json = try! encoder.encode(data)
      resolve(String(data: json, encoding: .utf8)!)
    }
  }

  // highlight-start
  @objc static override func requiresMainQueueSetup() -> Bool { return true }
  // highlight-end
}
```

## Add Your Form Component in Javascript

> Now it's time to mount `SsnTextElement` in Javascript. First, we'll go into the `App.tsx` file and setup the view where we'll display the secure text field.

Reviewer: ```suggestion
First, in `App.tsx` set up the view to display the secure text field.
```

/**
 * Sample React Native App
 * https://github.com/facebook/react-native
 *
 * @format
 */

import React from 'react';
// highlight-start
import {
  HostComponent,
  requireNativeComponent,
  SafeAreaView,
  StatusBar,
  StyleSheet,
  View,
} from 'react-native';

interface SsnTextElementProps {
  style: Record<string, unknown>;
}

const SsnTextElement: HostComponent<SsnTextElementProps> =
  requireNativeComponent('SsnTextElement');
// highlight-end

function App(): JSX.Element {
  return (
    <SafeAreaView style={styles.view}>
      <StatusBar />
      <View style={styles.container}>
        // highlight-start
        <SsnTextElement style={styles.ssnTextElement} />
        // highlight-end
      </View>
    </SafeAreaView>
  );
}

const styles = StyleSheet.create({
  view: {
    backgroundColor: '#f2f2f8',
  },
  container: {
    paddingHorizontal: 25,
    paddingTop: 75,
    height: '100%',
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'flex-start',
    justifyContent: 'flex-start',
  },
  // highlight-start
  ssnTextElement: {
    height: 50,
    width: '100%',
    marginBottom: 20,
  },
  // highlight-end
});

export default App;
```

> Lastly, we'll add a button to trigger our `tokenize` method and a text view to view our tokenize response.

Reviewer: the same concern here, and legit not sure what todo. This one sentence feels like it blends in between the code blocks. Maybe sub-steps if possible?

Reviewer: @brigonzalez I resolved that one because it had a suggestion - although do we have the option to make these substeps or do you not want to do that? 

this seems just a bit awkward of a transition to the next thing todo.

![image](https://user-images.githubusercontent.com/151590/222232550-dafc715d-e203-409b-94ed-a77f65cfd67d.png)

  );
}

const styles = StyleSheet.create({
  view: {
    backgroundColor: '#f2f2f8',
  },
  container: {
    paddingHorizontal: 25,
    paddingTop: 75,
    height: '100%',
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'flex-start',
    justifyContent: 'flex-start',
  },
  ssnTextElement: {
    height: 50,
    width: '100%',
    marginBottom: 20,
  },
  // highlight-start
  tokenize: {
    display: 'flex',
    width: 125,
    height: 50,
    border: 'none',
    backgroundColor: '#007aff',
    textAlign: 'center',
    alignItems: 'center',
    justifyContent: 'center',
    borderRadius: 16,
  },
  tokenizeText: {
    fontSize: 18,
    fontWeight: 'bold',
    color: 'white',
  },
  tokenText: {
    marginTop: 25,
  },
  // highlight-end
});

export default App;
```

🎉 The code above is the last bit that we need to tokenize! Now let's run the app by running `npx react-native run-ios` in your project. The screen should look something like this:

![](/img/iOS/ios-collect-data-iphone-screenshot.png)

## Conclusion

You can now capture any text data without your React Native app accessing the underlying value drastically reducing compliance and regulatory scope.

> Try typing in a social security number in the `ssnTextElement` and tap on Tokenize. Basis Theory's iOS SDK will pass the value for the element reference. This will
> be securely submitted directly to the [Create Token Endpoint](/docs/api/tokens#create-token). The resulting [token](/docs/concepts/what-are-tokens) is printed to the app.

Reviewer: ```suggestion
To test, type in a social security number in the `ssnTextElement` and tap on Tokenize. Basis Theory's iOS SDK securely submits the data directly to the [Create Token Endpoint](/docs/api/tokens#create-token). The resulting [token](/docs/concepts/what-are-tokens) is printed to the app.
```
